web/packages/shared/components/Validation/Validation.tsx

```diff
@@ -0,0 +1,190 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+
+import { Logger } from 'design/logger';
+
+import { isObject } from 'shared/utils/highbar';
+import { Store, useStore } from 'shared/libs/stores';
+
+import { ValidationResult } from './rules';
+
+const logger = new Logger('validation');
+
+/** A per-rule callback that will be executed during validation. */
+type RuleCallback = () => void;
+
+export type Result = ValidationResult | boolean;
+
+type ValidatorState = {
+  /** Indicates whether the last validation was successful. */
+  valid: boolean;
+  /**
+   * Indicates whether the validator has been activated by a call to
+   * `validate`.
+   */
+  validating: boolean;
+};
+
+/** A store that handles input validation and makes its results accessible. */
+export default class Validator extends Store<ValidatorState> {
+  state = {
+    valid: true,
+    validating: false,
+  };
+
+  /**
+   * @deprecated For temporary Enterprise compatibility only. Use {@link state}
+   * instead.
+   */
+  valid = true;
+
+  /** Callbacks that will be executed upon validation. */
+  private ruleCallbacks: RuleCallback[] = [];
+
+  /** Adds a rule callback that will be executed upon validation. */
+  addRuleCallback(cb: RuleCallback) {
+    this.ruleCallbacks.push(cb);
+  }
+
+  /** Removes a rule callback. */
+  removeRuleCallback(cb: RuleCallback) {
+    const index = this.ruleCallbacks.indexOf(cb);
+    if (index > -1) {
+      this.ruleCallbacks.splice(index, 1);
+    }
+  }
+
+  addResult(result: Result) {
+    // result can be a boolean value or an object
+    let isValid = false;
+    if (isObject(result)) {
+      isValid = result.valid;
+    } else {
+      logger.error(`rule should return a valid object`);
```

Comment:
Can we enforce it via types?

Reply:
This fragment is not touched from its existing implementation, and since it's called by an untyped code, it's hard to tell what's gonna break if I disallow using a plain boolean here. That's out of my scope.